pkg/relayer/proxy/relay_meter.go

```diff
@@ -129,39 +138,75 @@ func (rmtr *ProxyRelayMeter) Start(ctx context.Context) error {
 // The relay reward is added optimistically, assuming that the relay will be volume / reward
 // applicable and the relay meter would remain up to date.
 func (rmtr *ProxyRelayMeter) AccumulateRelayReward(ctx context.Context, reqMeta servicetypes.RelayRequestMetadata) error {
+	// TODO_MAINNET: Locking the relay serving flow to ensure that the relay meter is updated
+	// might be a bottleneck since ensureRequestAppMetrics is performing multiple
+	// sequential queries to the Pocket Network node.
+	// Re-evaluate when caching and invalidation is implemented.
 	rmtr.relayMeterMu.Lock()
 	defer rmtr.relayMeterMu.Unlock()
 
-	// Create a metric if it does not exist.
-	appMetrics, err := rmtr.ensureRequestAppMetrics(ctx, reqMeta, true)
+	// Ensure that the served application has a relay meter and update the consumed
+	// stake amount.
+	appRelayMeter, err := rmtr.ensureRequestSessionRelayMeter(ctx, reqMeta)
 	if err != nil {
 		return err
 	}
 
 	// Get the cost of the relay based on the service and shared parameters.
-	relayCostCoin, err := getMinedRelayCostCoin(
-		appMetrics.sharedParams,
-		appMetrics.service,
-		appMetrics.serviceRelayDifficulty,
+	relayCostCoin, err := getSingleMinedRelayCostCoin(
+		appRelayMeter.sharedParams,
+		appRelayMeter.service,
+		appRelayMeter.serviceRelayDifficulty,
 	)
 	if err != nil {
 		return err
 	}
 
 	// Increase the consumed stake amount by relay cost.
-	newConsumedCoin := appMetrics.consumedCoin.Add(relayCostCoin)
-
-	// If the consumed amount exceeds the max amount, return a rate limit error.
-	overServicingLimited := !rmtr.overServicingAllowanceCoins.IsNegative()
-	if overServicingLimited && appMetrics.maxCoin.IsLT(newConsumedCoin) {
-		return ErrRelayerProxyRateLimited.Wrapf(
-			"application has been rate limited, stake needed: %s, has: %s, ",
-			newConsumedCoin.String(),
-			appMetrics.maxCoin.String(),
+	newConsumedCoin := appRelayMeter.consumedCoin.Add(relayCostCoin)
+
+	isAppOverServiced := appRelayMeter.maxCoin.IsLT(newConsumedCoin)
+
+	if !isAppOverServiced {
+		appRelayMeter.consumedCoin = newConsumedCoin
+		return nil
+	}
+
+	// Check if the supplier is allowing unlimited over-servicing (i.e. negative value)
+	allowUnlimitedOverServicing := rmtr.overServicingAllowanceCoins.IsNegative()
+
+	// The application is over-servicing, if unlimited over-servicing is not allowed
+	// and the newConsumedCoin is greater than the maxCoin + overServicingAllowanceCoins,
+	// then return a rate limit error.
+	overServicingCoin := newConsumedCoin.Sub(appRelayMeter.maxCoin)
+
+	// In case Allowance is positive, add it to the maxCoin to allow no or limited over-servicing.
+	if !allowUnlimitedOverServicing {
+		maxAllowedOverServicing := appRelayMeter.maxCoin.Add(rmtr.overServicingAllowanceCoins)
```

Comment:
Could  (should?) this be a value that we cache on initialization or new sessions?

Reply:
There seems to be no need to caching or initialization, this is the only place where it's being used.

The `maxCoin` is updated at the beginning of each session, while `overServicingAllowanceCoins` are configured at startup and unchanged.